### PR TITLE
Run test file if it is changed or added

### PR DESCRIPTION
**Pull request by Genki-S**
Implementation of #23 

This PR makes it possible to select a test file as target if the test file itself is modified since the base (anchored) commit.

This might seem too big change for such a small change. The reason is that I needed to retrieve the instance of `TTNT::TestTask` in order to check if a given file name is a test file or not.

The second commit ( https://github.com/Genki-S/ttnt/commit/e5c1ba8381b07590e92414d43a8d8579b6f0d868 ) implements the check in simplest way, and the last commit ( https://github.com/Genki-S/ttnt/commit/fd6db90645c3eeb41f6c68a8804cf8e1ea837556 ) modifies the check so that it can pick any test file specified in `Rakefile`. All commits in between are the path to make the last change possible.

@robin850 could you review the code please?

### CODE REVIEW

**Comment by Genki-S** [comment]
(BTW I found many missing periods in documentation again... I'll fix this in another PR, sorry for that :sweat: )

File: lib/ttnt/test_selector.rb [hunk 1]
```diff
@@ -29,6 +32,8 @@ def select_tests
       diff = @base_obj.diff(@target_obj)
       diff.each_patch do |patch|
         file = patch.delta.old_file[:path]
+        tests << file and next if test_file?(file)
```
Comment (robin850) [inline lib/ttnt/test_selector.rb]: What do you think about making an explicit `if` / `else` like:

~~~ruby
if test_file?(file)
  tests << file
else
  patch.each_hunk do |hunk|
    hunk.each_line do |line|
      case line.line_origin
      when :addition
        tests += mapping.get_tests(file: file, lineno: line.new_lineno)
      when :deletion
        tests += mapping.get_tests(file: file, lineno: line.old_lineno)
      end
    end
  end
end
~~~

We easily see that we have two branches, one for each kind of files. Also feel free to ask me regarding your `FIXME` and `TODO` comments ; I do not understand them. :p

File: lib/ttnt/testtask.rb [hunk 2]
```diff
@@ -22,9 +30,19 @@ def initialize(rake_testtask)
...
+
+      @@instance_by_name[@rake_testtask.name.to_s] = self
       define_tasks
     end
 
+    # Returns array of test file names.
+    #   Unlike Rake::TestTask#file_list, patterns are expanded.
+    def expanded_file_list
+      test_files = Rake::FileList[@rake_testtask.pattern].compact
+      test_files += @test_files.to_a if @test_files
```
Comment (robin850) [inline lib/ttnt/testtask.rb]: Can't this method simply be defined as:

~~~ruby
def expanded_file_list
  Rake::FileList[@rake_testtask.pattern].compact
end
~~~

? At least the suite is passing with such method so this is either useless or you are missing a test where this line would be useful.

File: lib/ttnt/testtask.rb [hunk 3]
```diff
@@ -92,8 +110,7 @@ def define_anchor_task
             "-I#{gem_root} -r ttnt/anchor " +
             "#{@rake_testtask.ruby_opts_string}"
 
-          test_files = Rake::FileList[@rake_testtask.pattern].compact
-          test_files += @test_files.to_a if @test_files
+          test_files = expanded_file_list
```
Comment (robin850) [inline lib/ttnt/testtask.rb]: This assignment is useless ; you can simply call `expanded_file_list.each`.

File: test/test_selector_test.rb [hunk 4]
```diff
@@ -7,7 +7,8 @@ def setup
       target_sha = @repo.branches['change_fizz'].target.oid
       master_sha = @repo.branches['master'].target.oid
       base_sha = @repo.merge_base(target_sha, master_sha)
-      @selector = TTNT::TestSelector.new(@repo, target_sha, base_sha)
+      @testtask = TTNT::TestTask.get_instance('test')
```
Comment (robin850) [inline test/test_selector_test.rb]: Is the `#get_instance` method really useful ? Can't we just define a dummy `TTNT::TestTask` object here with a new `Rake::TestTask` like:

~~~ruby
rake_task = Rake::TestTask.new do |t|
  t.pattern = "test/**/*_test.rb"
end

@testtask = TTNT::TestTask.new(rake_task)
~~~

File: lib/ttnt/test_selector.rb (hunk 1 above)
Comment (Genki-S) [inline lib/ttnt/test_selector.rb]: Seems nice, thank you! I avoided it since logic in else block seemed a little too long to me. So, I made if-else explicit and moved the logic in else block to another method to make it more readable: https://github.com/Genki-S/ttnt/commit/a5f1fb61d74a114692a7b05f50feeeda86222a88

Comment (Genki-S) [inline lib/ttnt/test_selector.rb]: > Also feel free to ask me regarding your FIXME and TODO comments ; I do not understand them. :p

I'm sorry for unclear comments :sweat: I will check them out and ask you soon (maybe via another issue)

File: lib/ttnt/testtask.rb (hunk 2 above)
Comment (Genki-S) [inline lib/ttnt/testtask.rb]: We need to regard both `pattern` and `test_files` options specified for `Rake::TestTask` since it is possible that both options are specified by a user. (`Rake::TestTask#file_list` is [doing the same](https://github.com/ruby/rake/blob/master/lib/rake/testtask.rb#L146-L149))

> you are missing a test where this line would be useful

That's right! :bulb: Thank you for your suggestion :+1: I added it here: https://github.com/Genki-S/ttnt/commit/2ca95e7ca074562de55dc1d03e96c4f828b09749

File: lib/ttnt/testtask.rb (hunk 3 above)
Comment (Genki-S) [inline lib/ttnt/testtask.rb]: Oops, thank you for catching. Fixed here: https://github.com/Genki-S/ttnt/commit/e4a741a92bae5f08a60ad7cf9ca9e9d89d0b3f06

File: test/test_selector_test.rb (hunk 4 above)
Comment (Genki-S) [inline test/test_selector_test.rb]: That sounds better to me too... I did it that way since dummy `TTNT::TestTask` object defined that way can have different options from the one defined in the fixture `Rakefile`. But we need to keep the options consistent, and having unneeded method seems worse than managing consistency...

I reverted the commit which added `get_instance` (https://github.com/Genki-S/ttnt/commit/ab05776defffbc2567210f9484a464a8018d6960) and applied your approach to define a dummy `TTNT::TestTask` object (https://github.com/Genki-S/ttnt/commit/8070b7c3c23f1e4d03241f40abfbaa0c9d35c7b6).

(I will squash these commits just before merging)

File: lib/ttnt/test_selector.rb [hunk 5]
```diff
@@ -50,6 +42,33 @@ def select_tests
 
     private
 
+    def test_to_code_mapping
```
Comment (robin850) [inline lib/ttnt/test_selector.rb]: Since you are maintaining a `@mapping` instance variable, to be consistent and because we know which kind of mapping we are maintaining, you could simply rename it `mapping`. What do you think ?

File: test/test_selector_test.rb [hunk 6]
```diff
@@ -7,7 +7,16 @@ def setup
       target_sha = @repo.branches['change_fizz'].target.oid
       master_sha = @repo.branches['master'].target.oid
       base_sha = @repo.merge_base(target_sha, master_sha)
-      @selector = TTNT::TestSelector.new(@repo, target_sha, base_sha)
+
+      # NOTE: Use the same configuration for Rake::TestTask with fixtures/Rakefile
+      #       to create TTNT::TestTask which are consistent with actual data
+      rake_task = Rake::TestTask.new(:test) do |t|
+        t.libs << "test"
```
Comment (robin850) [inline test/test_selector_test.rb]: Do you really think that we should maintain a similar Rake task that the one in the fixture Rakefile ? I think we should only keep what's necessary for testing purpose ; there are edge cases anyway, the fixture file is just a general one.

File: test/testtask_test.rb [hunk 7]
```diff
@@ -25,4 +26,12 @@ def test_define_rake_tasks
   def test_composing_rake_testtask
     assert_equal @rake_task, @ttnt_task.rake_testtask
   end
+
+  def test_expanded_file_list
+    # It gathers tests from both `pattern` and `test_files` option for Rake::TestTask
+    assert_equal(
```
Comment (robin850) [inline test/testtask_test.rb]: You can certainly avoid the call on several lines with:

~~~ruby
test_files = Rake::FileList['test/**/*_test.rb'] + ['test/dummy_test.rb']
assert_equal test_files, @ttnt_task.expanded_file_list
~~~

**Comment by robin850** [comment]
Looking good!

A last comment regarding `select_tests` though. Instead of defining a `Set` object inside this method and one for each patch, you should rather do something like:

~~~ruby
attr_reader :tests

def initialize(repo, target_sha, base_sha, testtask)
  # ...
  @tests = Set.new
end

def select_tests
      # ...
      select_test_from_patch(patch)
    end
  end

  @tests.delete(nil)
end
~~~

and then access `@tests` instead of `tests` and remove the first and last line from `select_tests_from_patch`. Moreover, calling `#tests`, we can get the selected tests without having to walk through all the patch lines again. By the way, +1 for extracting the selection logic for patches to a method.

> I will squash these commits just before merging

:+1: for squashing these two commits.

Keep up the good work!

File: lib/ttnt/test_selector.rb (hunk 5 above)
Comment (Genki-S) [inline lib/ttnt/test_selector.rb]: You are right, we don't have other "mapping"s around. Renamed to `#mapping`: https://github.com/Genki-S/ttnt/commit/3a400be86b9e9693accec7fb7a1435bb3fc1804d

File: test/test_selector_test.rb (hunk 6 above)
Comment (Genki-S) [inline test/test_selector_test.rb]: I think so because `TTNT::TestSelector` expects to receive the same `TTNT::TestTask` object defined in `Rakefile`.

...and after saying that, I realized :bulb: that receiving an instance of `TTNT::TestTask` (and making such undocumented assumption) was really a bad design. So I changed `TTNT::TestSelector` to receive candidate test files rather than an instance of `TTNT::TestTask`. This way, `TTNT::TestSelector` can be instantiated with minimal data necessary for testing purpose.

Thank you for your critical comment which made me rethink about good design! :smile: 
Could you check if the approach is OK? Change is here: https://github.com/Genki-S/ttnt/commit/5b0eeb389e8d71c27818707854fac0a2ac02adab

File: test/testtask_test.rb (hunk 7 above)
Comment (Genki-S) [inline test/testtask_test.rb]: Cool, thank you! Fixed: https://github.com/Genki-S/ttnt/commit/8a0f17e6d6dd9912bea2c19e06e213d881c1f266

**Comment by Genki-S** [comment]
> A last comment regarding select_tests

> Moreover, calling #tests, we can get the selected tests without having to walk through all the patch lines again

Great! Thank you for giving me the idea :smile: I changed the implementation that way: https://github.com/Genki-S/ttnt/commit/3cb1dd2c53cbe33379bb66a40c8de744369db366 (I renamed `#select_tests` to `#select_tests!` since it changes the internal state)

One concern is that when `#tests` returns empty set, one cannot know if the cause is either (1) `#select_tests!` is not called yet or (2) `#select_tests!` is called, but no test is selected. I think it's OK, but it might be nicer if we provide a way to distinguish those (e.g. returning `nil` instead of empty set when calling `#tests` before calling `#select_tests!`. What do you think?

**Comment by robin850** [comment]
> I renamed #select_tests to #select_tests! since it changes the internal state

Yep, good point! :+1:

> One concern is that when #tests returns empty set, one cannot know if the cause is either (1) #select_tests! is not called yet or (2) #select_tests! is called, but no test is selected.

You could actually remove the `@tests = Set.new` line from `TTNT::TestSelector#initialize` and put `@tests ||= Set.new` inside `select_tests!`. Thus `@tests` will be nil if the latter isn't already called.

Once this is done you can squash the aforementioned commits and merge this! :-)

**Comment by Genki-S** [comment]
> You could actually remove the @tests = Set.new line from TTNT::TestSelector#initialize and put @tests ||= Set.new inside select_tests!

Cool! Thank you :smile: Done here: https://github.com/Genki-S/ttnt/commit/ddbd5c9117b2f8575a2f5eb284e273f01367b230

> Once this is done you can squash the aforementioned commits and merge this! :-)

Thank you so much for your comments! This PR got way better than the beginning :dancer: 
I'll tidy up the history and merge this :+1:

**Comment by robin850** [comment]
Awesome! Good work! :-)
